nmigen/build/run.py

```diff
@@ -99,6 +100,64 @@ def execute_local(self, root="build", *, run_script=True):
         finally:
             os.chdir(cwd)
 
+    def execute_remote_ssh(self, root, hostname, *, connect_args = {}):
+        """
+        Execute build plan using the remote SSH strategy. Files from the build
+        plan are transferred via SFTP to the directory ``root`` on the remote
+        server ``hostname``. The ``paramiko`` SSH client will then run ``{script}.sh``.
+
+        ``hostname`` corresponds to the first (required) input argument of ``paramiko``'s
+        ``SSHClient.connect``, and ``connect_args`` is a dictionary that holds
+        all other input arguments to ``SSHClient.connect``
+        (`documentation <http://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.connect>`_).
+
+        Returns :class:`RemoteSshBuildProducts`.
+        """
+        from paramiko import SSHClient
+
+        with SSHClient() as client:
+            client.load_system_host_keys()
+            client.connect(hostname, **connect_args)
+
+            with client.open_sftp() as sftp:
+                try:
+                    sftp.mkdir(root)
+                except IOError as e:
+                    pass # mkdir fails if directory exists. This is fine.
+
+                sftp.chdir(root)
+                for filename, content in self.files.items():
+                    filename = os.path.normpath(filename)
+                    # Just to make sure we don't accidentally overwrite anything outside of build root.
+                    assert not filename.startswith("..")
+
+                    dirname = os.path.dirname(filename)
+                    if dirname:
+                        try:
+                            sftp.mkdir(dirname, exist_ok=True)
+                        except IOError as e:
+                            pass
+
+                    mode = "wt" if isinstance(content, str) else "wb"
+                    with sftp.file(filename, mode) as f:
+                        f.write(content)
+
+            cmd = "cd {} && bash -l {}.sh".format(root, self.script)
```

Comment:
You shouldn't call bash directly. Not all build scripts even use bash.

Reply:
You are correct, my mistake.

`-l` argument is a bash-ism (TIL!). `dash` also has `-l`, but other variants don't. So `sh -l` won't work in general to invoke a login shell. Since all we really want is to source `.profile`<sup>1</sup>, what about:

`if [ -f ~/.profile ]; then . ~/.profile; fi && cd {} && sh {}.sh`

1. Even if `sh` is a symlink to `bash`, `.bash_profile` will [not be read](https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html) if invoked via the name `sh`.

Reply:
Based on the doc you linked, it seems that the most portable option would be to run `cd {root} && sh {self.script}`, and allow setting arbitrary environment variables, which may include `ENV`, `NMIGEN_ENV_<toolchain>`, or just `<tool_name>` directly depending on the use case.